### ISSUE TITLE
GeneralizedRCNNTransform output different shapes of images and targets

### ISSUE THREAD

**Issue by lpq97lpq** [pytorch/vision#6213]
### 🐛 Describe the bug

I would expect the shape of images and targets consistent with each other (that is, have the same width and height). 

However, since torchvision.models.detection.transform.GeneralizedRCNNTransform call self.batch_images at the end of its forward function only on images, the output images and targets usually don't have the same shapes.

Example:
```
from torchvision.models.detection.transform import GeneralizedRCNNTransform
import torch

#initialize transormation
image_mean = [0.485, 0.456, 0.406]
image_std = [0.229, 0.224, 0.225]
T=GeneralizedRCNNTransform(800,1333,image_mean,image_std)

#initialize image & target
images=[torch.randn([3,1000,900])]
targets=[{'boxes':torch.tensor([[0,0,1000,900]]),'masks':torch.randn([1,1000,900]).byte()}]

#do experiment
images,targets=T(images,targets)

print(images.tensors.shape) 
print(images.image_sizes)
print(targets[0]['boxes'])
print(targets[0]['masks'].shape)

```

output:

torch.Size([1, 3, 896, 800])
[(888, 800)]
tensor([[  0.0000,   0.0000, 888.8889, 799.2000]])
torch.Size([1, 888, 800])

### Versions

Collecting environment information...
PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.8 (default, Apr 13 2021, 19:58:26)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.11.0-37-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.4.120
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 2070
Nvidia driver version: 470.57.02
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.20.1
[pip3] numpydoc==1.1.0
[pip3] pytorch-lightning==1.3.3
[pip3] torch==1.11.0
[pip3] torch-summary==1.4.5
[pip3] torchaudio==0.11.0
[pip3] torchmetrics==0.5.1
[pip3] torchvision==0.12.0
[pip3] vit-pytorch==0.22.0
[conda] blas                      1.0                         mkl  
[conda] cudatoolkit               11.3.1               h2bc3f7f_2  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2021.2.0           h06a4308_296  
[conda] mkl-service               2.3.0            py38h27cfd23_1  
[conda] mkl_fft                   1.3.0            py38h42c9631_2  
[conda] mkl_random                1.2.1            py38ha9443f7_2  
[conda] numpy                     1.20.1           py38h93e21f0_0  
[conda] numpy-base                1.20.1           py38h7d8b39e_0  
[conda] numpydoc                  1.1.0              pyhd3eb1b0_1  
[conda] pytorch                   1.11.0          py3.8_cuda11.3_cudnn8.2.0_0    pytorch
[conda] pytorch-lightning         1.3.3                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torch                     1.11.0                   pypi_0    pypi
[conda] torch-summary             1.4.5                    pypi_0    pypi
[conda] torchaudio                0.11.0               py38_cu113    pytorch
[conda] torchmetrics              0.5.1                    pypi_0    pypi
[conda] torchvision               0.12.0                   pypi_0    pypi
[conda] vit-pytorch               0.22.0                   pypi_0    pypi

**Comment by datumbox** [comment]
@lpq97lpq Thanks for raising this. I agree, this is very confusing.

The `GeneralizedRCNNTransform` is not meant to be used directly by users (though that's not really obvious). In the current Detection API (which is still Beta), the models include the transforms internally. This was done so that they can undo internal resizing ops such as the ones you point out. Here is the intended usage:

```python
import torch

images=[torch.randn([3,1000,900])]
from torchvision.models.detection import maskrcnn_resnet50_fpn_v2

model = maskrcnn_resnet50_fpn_v2().eval()
predictions = model(images)


print(images[0].shape)
print(predictions[0]["masks"].shape)
```

The above prints:
```
torch.Size([3, 1000, 900])
torch.Size([100, 1, 1000, 900])
```

This means that the resizing done by `size_divisible` is not visible to the users. 

The API of Detection models has multiple pros/cons. Here are the pros:
- The model can do all sorts of internal optimizations and these are transparent to the user
- The model is capable to perform internally postprocessing (such as NMS) and resizing and return back to the user the expected results.
- The API can support 2-stage models where different components have different losses without the user needing to worry about them.

But there are also cons:
- The detection models look nothing like the standard Model paradigm in PyTorch. The losses are parts of the model, the forward behaves differently depending on whether we do train/eval, the images are transformed within the model instead of outside.
- There is limitation on the amount of data augmentation strategies we can do. Overtime we had to introduce hacky and ugly patches to make the use of modern augmentations possible (see this workaround for which I'm not too proud: https://github.com/pytorch/vision/pull/5715/files#r839744536)

The Detection API is in need of a revamp. It will require careful thinking on how this will be done because one can't optimize for all use-cases. One direction we consider taking is to make detection models compatible with the standard paradigm of Models in PyTorch. That would include moving Transforms and Losses out of the model and requiring the user to explicitly do the conversions themselves. Unfortunately this would also mean handling 2-stage models explicitly by calling their components and losses directly; that is far from ideal. We haven't finalized the plan on this and currently we are running low on headcount to prioritize this work. But I will leave your issue open as a great example of the things we need to improve concerning the API. Thank you for taking the time to report the issue. Let me know your thoughts.

**Comment by fmassa** [comment]
Hi,

Thanks for raising this question. There is a slight misunderstanding in one point in here though.

`batch_images` adds padding on the bottom-right side of the images (if needed). Note that the bounding boxes are referenced starting from the top-left corner of the image, in absolute coordinates.

This means that indeed performing the batching as it was implemented doesn't affect the coordinates of the boxes.

Note that this wouldn't hold true if the boxes were represented as relative coordinates (i.e., in 0-1), because the padding done in the images would affect the value required in the boxes for them to represent the same value (e.g., 1.0 in relative coordinates for an image of size 100, if padded to size 200, would correspond to 0.5).

The slight difference in values of doing padding is due to the convolutions in the backbone having slightly different results due to padding, but this is very small as a difference and shouldn't matter much in practice

I hope this answers the question.
